### PR TITLE
cOmx: Correct and optimize the mutex usage

### PR DESCRIPTION
**Pull request by dr-m**
`cOmx::HandlePortBufferEmptied()`: Determine the address of the variable before acquiring the mutex. The addresses will not change while they are accessible to multiple threads.

`cOmx::HandlePortSettingsChanged()`: Only `Lock()` around changes of `cOmx` data members. This function is only being called by the `cOmx::Action()` thread, so the IL calls should not require additional mutex protection.

`cOmx::EmptyAudioBuffer()`, `cOmx::EmptyVideoBuffer()`: Determine the return value outside the critical section.

### CODE REVIEW

**Comment by dr-m** [comment]
Please verify my claim about `cOmx::HandlePortSettingsChanged()`.

Note: I also attempted to replace `cOmx::m_mutex` with `cThread::mutex` (after declaring the latter `protected`), but that would lead to hangs. I am a little uncertain whether all calls to `Lock()` and `Unlock()` in `omx.c` are really necessary at all. It could be helpful to document each data member: whether they are private to the `cOmx::Action()` thread, or which mutex is protecting them.

**Comment by dr-m** [comment]
If I proceed further to replace any `Lock()` in `cOmx` with `m_mutex.Lock()`, then the playback will hang, at least some of the time.

I wonder if it would be possible to make all the member functions that currently acquire `cOmx::cThread::mutex` to use the `cOmx::Add()` work queue. In that way, all the `ilclient_` and `OMX_` functions would be invoked either by `cOmx::Init()` before the thread for running `cOmx::Action()` is started, or by `cOmx::Action()`.

**Comment by dr-m** [comment]
I developed further my idea of making use of the `cOmx::Add()` event queue as a work queue, in dr-m/rpihddevice@dbd4f8d7c7b310c48d5d7419115a759e17f2e52c. It seems to work. I believe that `cOmx::SetupAudioRender()` could be refactored in the same style (making the `cOmx::cEvent::data` a pointer), but I did not try it yet.

I did not check `cOmx::GetAudioBuffer()`, `cOmx::GetVideoBuffer()`, `cOmx::EmptyAudioBuffer()`, `cOmx::EmptyVideoBuffer()` yet, but I have the feeling that they may need to be processed ahead of other requests in the `m_portEvents` queue. I suppose that they could be added to their own work queue that would be prioritized over `m_portEvents` by `cOmx::Action()`.

**Comment by dr-m** [comment]
After experimenting a little, I concluded that we need both `cOmx::mutex` and `cOmx::m_mutex`. The latter should protect the event queue, while `cOmx::mutex` needs to protect the `ilclient_` and `OMX_` calls as well as all data members that are related to the buffer management. At least `OMX_EmptyThisBuffer()` will wait for some `cOmx::m_portEvents` processing from another thread. That processing is protected by `cOmx::m_mutex`. If the caller of `OMX_EmptyThisBuffer()` is holding `cOmx::m_mutex`, a deadlock will occur.

I tried to move most commands to be executed via `cOmx::m_portEvents` by the `cOmx::Action()`, but that does not seem to offer any benefit. I observed a disadvantage: jerky action when switching from normal playback speed to fast-forward.

**Comment by dr-m** [comment]
I tried to remove `cOmx::m_portEvents` and just let the callback functions invoke the handling actions directly. I did not figure out how to avoid a hang or deadlock in `cOmx::HandlePortSettingsChanged()`, even when retaining 2 mutexes in  `cOmx` (one to protect the entry to `OMX_` and `ilclient_` functions, and another one to protect data members in the callback functions).

It might be possible to bypass the event queue for `cOmx::OnEndOfStream()` and `cOmx::HandlePortBufferEmptied()`, but I did not test that, nor read any documentation to determine whether the resulting out-of-order processing could cause trouble.

In other words, I don’t think that I can improve `cOmx` beyond 099458c94f1c500387f34d42773e5caeaf5d4965.
